### PR TITLE
Add checks for completeness of msg_handler

### PR DESCRIPTION
**Pull request by rayrayraykk**
# Main changes
Add a check for completeness of msg_handler in workers. (Required `networkx` and `matplotlib` and I add them in `dev` mode)
## What's New in worker
A `classmethod`: `get_msg_handler_dict`;
A `self.msg_handlers_str` to store triplets <`msg_in`, `handler_name`, `[msg_out0, msg_out1, ...]`>

## How to use:
Set `cfg.check_completeness=True`:
```
python federatedscope/main.py --cfg scripts/example_configs/femnist.yaml check_completeness True
```

## What information will the user know:
If the checks are enabled, three types of info will show: `pass`, `warning`, `error` for the case:
completeness

And a DiGraph will be plotted in `exp_dir` to remind users something might goes wrong:
image

image

For pure FedAvg:
image

### CODE REVIEW

**Comment by rayrayraykk** [comment]
@xieyxclack Please help me check the `out_msg` type in `_register_default_handlers`, thanks!

File: federatedscope/core/fed_runner.py [hunk 1]
```diff
@@ -427,3 +432,62 @@ def _handle_msg(self, msg, rcv=-1):
...
                     download_bytes)
+
+    def check(self):
+        """
+        Check the completeness of Server and Client.
+
+        Returns:
+
+        """
+        if self.cfg.check_completeness:
```
Comment (joneswong) [inline federatedscope/core/fed_runner.py:443]: I don't think it is necessary to indent such a huge block of code

File: federatedscope/core/fed_runner.py [hunk 2]
```diff
@@ -427,3 +432,62 @@ def _handle_msg(self, msg, rcv=-1):
...
+                            f'Completeness check raises warning for '
+                            f'some handlers not in FL process! Save '
+                            f'check results in {fig_path}.')
+                else:
+                    logger.error(f'Completeness check fails for there is no'
+                                 f'path from `join_in` to `finish`! Save '
+                                 f'check results in {fig_path}.')
+            except Exception as error:
+                logger.warning(f'Completeness check failed for {error}!')
+        return
```
Comment (joneswong) [inline federatedscope/core/fed_runner.py:493]: so the reason is that we cannot say yes or not about the correctness/completeness, right?

Comment (rayrayraykk) [reply federatedscope/core/fed_runner.py:493]: As the document shows, the correctness/completeness checks only check the message-handler pairs and raise three types of logs (info, warning, and error). If something goes wrong with Python code, we'd better keep the exception stack as it is. So the return value is meaningless.

File: federatedscope/core/fed_runner.py (hunk 1 above)
Comment (rayrayraykk) [reply federatedscope/core/fed_runner.py]: Update accordingly.

**Review by xieyxclack — APPROVED** [review]
LGTM

**Comment by rayrayraykk** [comment]
DO NOT merge this PR ahead of #365 .

**Review by xieyxclack — APPROVED** [review]

**Pull request merged**